File: src/main/java/com/github/sorusclient/client/ui/framework/constraint/Dependent.java
package com.github.sorusclient.client.ui.framework.constraint;

import com.github.sorusclient.client.ui.framework.Component;
import com.github.sorusclient.client.util.Color;

import java.util.Map;
import java.util.function.Function;

public class Dependent implements Constraint {

    private final Function<Map<String, Object>, Object> function;

    public Dependent(Function<Map<String, Object>, Object> function) {
        this.function = function;
    }

    @Override
    public double getXValue(Component.Runtime componentRuntime) {
        Object value = this.function.apply(componentRuntime.getAvailableState());
        if (value instanceof Constraint) {
            return ((Constraint) value).getXValue(componentRuntime);
        } else {
            return (double) value;
        }
    }

    @Override
    public double getYValue(Component.Runtime componentRuntime) {
        Object value = this.function.apply(componentRuntime.getAvailableState());
        if (value instanceof Constraint) {
            return ((Constraint) value).getYValue(componentRuntime);
        } else {
            return (double) value;
        }
    }

    @Override
    public double getWidthValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public double getHeightValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public double getCornerRadiusValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public double getPaddingValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public Color getColorValue(Component.Runtime componentRuntime) {
        return (Color) this.apply(componentRuntime);
    }

    @Override
    public String getStringValue(Component.Runtime componentRuntime) {
        return null;
    }

    private Object apply(Component.Runtime componentRuntime) {
        return this.function.apply(componentRuntime.getAvailableState());
    }

}

File: src/main/kotlin/com/github/sorusclient/client/ui/framework/constraint/Dependent.kt
package com.github.sorusclient.client.ui.framework.constraint

import com.github.sorusclient.client.ui.framework.Component
import com.github.sorusclient.client.util.Color
import java.util.function.Function

class Dependent(private val function: Function<Map<String, Any>, Any>) : Constraint {
    override fun getXValue(componentRuntime: Component.Runtime): Double {
        val value = function.apply(componentRuntime.availableState)
        return if (value is Constraint) {
            value.getXValue(componentRuntime)
        } else {
            value as Double
        }
    }

    override fun getYValue(componentRuntime: Component.Runtime): Double {
        val value = function.apply(componentRuntime.availableState)
        return if (value is Constraint) {
            value.getYValue(componentRuntime)
        } else {
            value as Double
        }
    }

    override fun getWidthValue(componentRuntime: Component.Runtime): Double {
        val value = function.apply(componentRuntime.availableState)
        return if (value is Constraint) {
            value.getWidthValue(componentRuntime)
        } else {
            value as Double
        }
    }

    override fun getHeightValue(componentRuntime: Component.Runtime): Double {
        val value = function.apply(componentRuntime.availableState)
        return if (value is Constraint) {
            value.getHeightValue(componentRuntime)
        } else {
            value as Double
        }
    }

    override fun getCornerRadiusValue(componentRuntime: Component.Runtime): Double {
        return 0.0
    }

    override fun getPaddingValue(componentRuntime: Component.Runtime): Double {
        return 0.0
    }

    override fun getColorValue(componentRuntime: Component.Runtime): Color {
        val value = function.apply(componentRuntime.availableState)
        return if (value is Constraint) {
            value.getColorValue(componentRuntime)
        } else {
            value as Color
        }
    }

    override fun getStringValue(componentRuntime: Component.Runtime?): String {
        return this.apply(componentRuntime) as String
    }

    private fun apply(componentRuntime: Component.Runtime?): Any {
        return function.apply(componentRuntime?.availableState!!)
    }
}